### PR TITLE
Add structured config

### PR DESCRIPTION
**Pull request by welpaolo**
This PR includes structured config for the Kafka K8s charm.

### CODE REVIEW

File: src/structured_config.py [hunk 1]
```diff
@@ -0,0 +1,188 @@
...
+        if int_value < 1:
+            raise ValueError("Value below 1. Accepted value are greater or equal than 1.")
+        return int_value
+
+    @validator("replication_quota_window_num", "log_segment_bytes", "message_max_bytes")
+    @classmethod
+    def greater_than_zero(cls, value: int) -> Optional[int]:
+        """Check value greater than zero."""
+        if value < 0:
+            raise ValueError("Value below -1. Accepted value are greater or equal than -1.")
```
Comment (deusebio) [inline src/structured_config.py:144]: Probably Rob already mentioned this, but this should probably be 0

File: src/structured_config.py [hunk 2]
```diff
@@ -0,0 +1,188 @@
...
+    LZ4 = "lz4"
+    ZSTD = "zstd"
+    UNCOMPRESSED = "uncompressed"
+    PRODUCER = "producer"
+
+
+class CharmConfig(BaseConfigModel):
+    """Manager for the structured configuration."""
+
+    compression_type: str
```
Comment (deusebio) [inline src/structured_config.py:44]: **question** But the type here...shouldn't this be CompressionType? And similar for other enum types?

File: src/structured_config.py [hunk 3]
```diff
@@ -0,0 +1,200 @@
...
+    @classmethod
+    def greater_than_zero(cls, value: int) -> Optional[int]:
+        """Check value greater than zero."""
+        if value < 0:
+            raise ValueError("Value below 0. Accepted values are greater or equal than 0.")
+        return value
+
+    @validator("compression_type")
+    @classmethod
+    def value_compression_type(cls, value: str) -> Optional[str]:
```
Comment (deusebio) [inline src/structured_config.py:161]: I thought that this was not needed. Was I wrong?

File: src/structured_config.py [hunk 4]
```diff
@@ -0,0 +1,200 @@
...
+        if value == "":
+            return None
+        return value
+
+    @validator("log_message_timestamp_type")
+    @classmethod
+    def log_message_timestamp_type_validator(cls, value: str) -> Optional[str]:
+        """Check validity of `log_message_timestamp_type` field."""
+        try:
+            _log_message_timestap_type = LogMessageTimestampType(value)
```
Comment (deusebio) [inline src/structured_config.py:88]: Is this needed?

File: src/structured_config.py [hunk 5]
```diff
@@ -0,0 +1,200 @@
...
+        return str(self.value)
+
+    COMPACT = "compact"
+    DELETE = "delete"
+
+
+class CompressionType(str, Enum):
+    """Enum for the `compression_type` field."""
+
+    def __str__(self) -> str:
```
Comment (deusebio) [inline src/structured_config.py:41]: Maybe one could factor this out in a general class

```|
Class BaseStrEnum(str,Enum):
    Def __str__(self):
        ....

```

**Review by deusebio — APPROVED** [review]
All good! Just some suggestions

File: src/structured_config.py (hunk 3 above)
Comment (welpaolo) [reply src/structured_config.py]: Removed.

File: src/structured_config.py (hunk 4 above)
Comment (welpaolo) [reply src/structured_config.py]: Removed

File: src/structured_config.py (hunk 5 above)
Comment (welpaolo) [reply src/structured_config.py]: Done

File: src/config.py [hunk 6]
```diff
@@ -27,7 +27,7 @@ class KafkaConfig:
 
     def __init__(self, charm):
         self.charm = charm
-        self.default_config_path = f"{self.charm.config['data-dir']}/config"
+        self.default_config_path = f"{DATA_DIR}/config"
```
Comment (zmraul) [inline src/config.py:30]: should `/config` be appended to `DATA_DIR` literal directly?

File: tests/unit/test_structured_config.py [hunk 7]
```diff
@@ -0,0 +1,169 @@
+#!/usr/bin/env python3
```
Comment (zmraul) [inline tests/unit/test_structured_config.py:1]: **praise**: These are like the quintessential unit tests :)

**Review by zmraul — APPROVED** [review]
It's will be really good to have checks on config options from now on. Great PR!

File: src/config.py (hunk 6 above)
Comment (welpaolo) [reply src/config.py:30]: This is what was done before. I would keep it like this for the moment.

**Pull request merged**